src/action.c

```diff
@@ -103,6 +104,26 @@ int main(int argc, char **argv) {
 
     singularity_priv_drop_perm();
 
+    if ( singularity_registry_get("DAEMON_JOIN") == NULL  &&
+         singularity_registry_get("PIDNS_ENABLED") != NULL &&
+         singularity_registry_get("NOSHIMINIT") == NULL ) {
+
+        // At this point, we are now PID 1; when we later exec the payload,
+        // it will also be PID 1.  Unfortunately, PID 1 in Linux has special
+        // signal handling rules (the _only_ signal that will terminate the
+        // process is SIGKILL; all other signals are ignored).  Hence, we fork
+        // one more time.  This makes PID 1 a shim process and the payload
+        // process PID 2 (meaning that the payload gets the "normal" signal
+        // handling rules it would expect).
+
+        // Set the name of the process for ps
+        prctl(PR_SET_NAME, "shim-init", 0, 0, 0);
+        // And for ps -f
+        strncpy(argv[0], "shim-init", strlen(argv[0]));
```

Comment:
Overflow is possible there, we can't be sure there is enough room for argv[0] since it can be set by user and contain only 1 character